pkg/cmd/init.go

```diff
@@ -41,112 +55,205 @@ func loadCA(cfg *config.MicroshiftConfig) error {
-		return err
+		return nil, err
 	}
 
 	// ocp
 	if err := util.GenCerts("openshift-controller-manager", filepath.Join(cfg.DataDir, "/resources/openshift-controller-manager/secrets"),
 		"tls.crt", "tls.key",
 		[]string{"openshift-controller-manager", cfg.NodeName, cfg.NodeIP, "127.0.0.1", "kubernetes.default.svc", "kubernetes.default",
 			"kubernetes", "localhost"}); err != nil {
-		return err
+		return nil, err
 	}
 	if err := util.GenCerts("service-ca", filepath.Join(cfg.DataDir, "/resources/service-ca/secrets/service-ca"),
 		"tls.crt", "tls.key",
 		[]string{"localhost", cfg.NodeIP, "127.0.0.1", cfg.NodeName, apiServerServiceIP.String()}); err != nil {
-		return err
+		return nil, err
 	}
-	return nil
+	return certConfigs, nil
 }
 
-func initKubeconfig(cfg *config.MicroshiftConfig) error {
+func initKubeconfig(
+	cfg *config.MicroshiftConfig,
+	certConfigs *tlsConfigs,
+) error {
+	clusterTrustBundlePEM := certConfigs.clusterTrustBundle
+
 	if err := util.Kubeconfig(filepath.Join(cfg.DataDir, "/resources/kubeadmin/kubeconfig"),
+		clusterTrustBundlePEM,
 		"system:admin", []string{"system:masters"}, cfg.Cluster.URL); err != nil {
 		return err
 	}
 	if err := util.Kubeconfig(filepath.Join(cfg.DataDir, "/resources/kube-apiserver/kubeconfig"),
+		clusterTrustBundlePEM,
 		"kube-apiserver", []string{"kube-apiserver", "system:kube-apiserver", "system:masters"}, cfg.Cluster.URL); err != nil {
 		return err
 	}
-	if err := util.Kubeconfig(filepath.Join(cfg.DataDir, "/resources/kube-controller-manager/kubeconfig"),
-		"system:kube-controller-manager", []string{"system:kube-controller-manager"}, cfg.Cluster.URL); err != nil {
+
+	kcmCertPEM, kcmKeyPEM, err := certConfigs.kubeControllerManager.GetPEMBytes()
+	if err != nil {
+		return err
+	}
+	if err := util.KubeConfigWithClientCerts(
+		filepath.Join(cfg.DataDir, "resources", "kube-controller-manager", "kubeconfig"),
+		cfg.Cluster.URL,
+		clusterTrustBundlePEM,
+		kcmCertPEM,
+		kcmKeyPEM,
+	); err != nil {
+		return err
+	}
+
+	schedulerCertPEM, schedulerKeyPEM, err := certConfigs.kubeScheduler.GetPEMBytes()
+	if err != nil {
 		return err
 	}
-	if err := util.Kubeconfig(filepath.Join(cfg.DataDir, "/resources/kube-scheduler/kubeconfig"),
-		"system:kube-scheduler", []string{"system:kube-scheduler"}, cfg.Cluster.URL); err != nil {
+	if err := util.KubeConfigWithClientCerts(
+		filepath.Join(cfg.DataDir, "resources", "kube-scheduler", "kubeconfig"),
+		cfg.Cluster.URL,
+		clusterTrustBundlePEM,
+		schedulerCertPEM, schedulerKeyPEM,
+	); err != nil {
 		return err
 	}
+
 	// per https://kubernetes.io/docs/reference/access-authn-authz/node/#overview
 	if err := util.Kubeconfig(filepath.Join(cfg.DataDir, "/resources/kubelet/kubeconfig"),
+		clusterTrustBundlePEM,
 		"system:node:"+cfg.NodeName, []string{"system:nodes"}, cfg.Cluster.URL); err != nil {
 		return err
 	}
 	return nil
 }
+
+func generateClientCA(certsDir, signerDir, signerName string) (*crypto.CA, error) {
```

Comment:
Shouldn't this be in cryptomaterial package or util?

Reply:
For now I treated it only as a shortcut function that is very specific to what's done in this file. Would it be useful elsewhere?

Reply:
yeah, if this is not gonna be reuse, just keep it as a local function. I'm ok with it.